addons/l10n_hk/models/res_bank.py

```diff
@@ -10,50 +10,64 @@
+        bank_hk = self.filtered(lambda b: b.country_code == 'HK')
+        bank_hk.display_qr_setting = True
+        super(ResPartnerBank, self - bank_hk)._compute_display_qr_setting()
+
     # Follow the documentation of FPS QR Code Standard [1]
     # [1]: https://www.hkma.gov.hk/media/eng/doc/key-functions/financial-infrastructure/infrastructure/retail-payment-initiatives/Common_QR_Code_Specification.pdf
     def _get_merchant_account_info(self):
         if self.country_code == 'HK':
             fps_type_mapping = {
-                'id': '02',
-                'mobile': '03',
-                'email': '04'
+                'id': 2,
+                'mobile': 3,
+                'email': 4,
             }
-            fps_type_id = fps_type_mapping[self.l10n_hk_fps_type]
-            fps_account = f'{fps_type_id}{len(self.l10n_hk_fps_identifier):02}{self.l10n_hk_fps_identifier}'
-            return f'0012hk.com.hkicl{fps_account}'
+            fps_type = fps_type_mapping[self.proxy_type]
+            merchant_account_vals = [
+                (0, 'hk.com.hkicl'),                                            # GUID
+                (fps_type, self.proxy_value),                        # Proxy Type and Proxy Value
+            ]
+            merchant_account_info = ''.join([self._serialize(*val) for val in merchant_account_vals])
+            return merchant_account_info
         return super()._get_merchant_account_info()
 
     def _get_additional_data_field(self, comment):
-        return f'05{len(comment):02}{comment}'
+        if self.country_code == 'HK':
+            return self._serialize(5, comment)
+        return super()._get_additional_data_field(comment)
 
     def _get_error_messages_for_qr(self, qr_method, debtor_partner, currency):
         if qr_method == 'emv_qr' and self.country_code == 'HK':
             if currency.name not in ['HKD', 'CNY']:
-                return _("Can't generate a FPS QR code with currency other than HKD or CNY.")
+                return _("Can't generate a FPS QR code with a currency other than HKD or CNY.")
             return None
 
         return super()._get_error_messages_for_qr(qr_method, debtor_partner, currency)
 
     def _check_for_qr_code_errors(self, qr_method, amount, currency, debtor_partner, free_communication, structured_communication):
-        if qr_method == 'emv_qr' and self.country_code == 'HK' and (not self.l10n_hk_fps_type or not self.l10n_hk_fps_identifier):
-            return _("The account receiving the payment must have a FPS type and a FPS identifier set.")
+        if qr_method == 'emv_qr' and self.country_code == 'HK':
+            if not self.proxy_type or not self.proxy_value:
+                return _("The account receiving the payment must have a FPS type and a FPS identifier set.")
+            if self.proxy_type not in ['id', 'mobile', 'email']:
+                return _("The account receiving the payment must have a valid FPS type set.")
```

Comment:
I think it would be clearer for the user if the message lists the correct values.
Something like:
```suggestion
                return _("""The account receiving the payment must have a valid FPS type set.
                              The supported types are ...""")
```
I think it could also be better to generalize the _check_for_qr_code_errors as it is going to be the same every time:
- Check that both values are filled
- Check that the proxy_type is correct

You could just put it in account_qr_code_emv and have a method that returns the available types based on the country that you would override. This is just a suggestion.

Reply:
I put the `Check that both values are filled` into account_qr_code_emv. Seems better to put in this way